### PR TITLE
Support inference for SPRL(Self -Play Reinforcement Learning).

### PR DESCRIPTION
**Pull request by adoda**
Support an simple demo for SPRL(Self -Play Reinforcement Learning)

### CODE REVIEW

**Review by haolin-nju — COMMENTED** [review]
See comments. BTW, what's the difference between `SPRLEnv` and `MCTSEnv`, since they look quite similar in code?

**Comment by adoda** [comment]
> See comments. BTW, what's the difference between `SPRLEnv` and `MCTSEnv`, since they look quite similar in code?

There are some similarities in the process, both being directed acyclic graphs (DAGs), but there are differences in state management and execution paths. Later, we can consider extracting the execution logic of the DAG.

File: chatlearn/runtime/environment.py [hunk 1]
```diff
@@ -213,4 +213,68 @@ def execute(self, is_eval):
             self.get_all_merged_data(data, out_queue, encode=False)
         return out_queue
 
+class SPRLEnv(Environment):
+    """SPRL Env"""
```
[the remark this answers is not in the document]
Comment (adoda) [reply chatlearn/runtime/environment.py]: done

File: chatlearn/runtime/environment.py [hunk 2]
```diff
@@ -213,4 +213,68 @@ def execute(self, is_eval):
...
+            current_data_producer = next(data_producer_iter)
+            query = current_data_producer.master.next_batch.remote(is_eval=is_eval)
+            encoded_data = encode_data(mb, query)
+            replica_data_list = []
+            model_to_replica = {}
+            for model_group in self.model_flow.flow_topology:
+                for model_node in model_group:
+                    model = model_node.model
+                    assert not model.is_colocate, "colocation is currently not supported in MCTSEnv"
+                    assert not model.enable_offload, "offload is currently not supported in MCTSEnv"
```
[the remark this answers is not in the document]
Comment (adoda) [reply chatlearn/runtime/environment.py]: done

**Review by haolin-nju — APPROVED** [review]
LGTM

**Review by charles9304 — APPROVED** [review]

**Pull request merged**